java/arcs/core/allocator/Allocator.kt

```diff
@@ -133,8 +126,6 @@ class Allocator(val hostRegistry: HostRegistry) {
      */
     private suspend fun findArcHostBySpec(spec: ParticleSpec): ArcHost =
         hostRegistry.availableArcHosts()
-            .filter { host ->
-                host.registeredParticles().map { clazz -> clazz.java.getCanonicalName() }
-                    .contains(spec.location)
-            }.firstOrNull() ?: throw ParticleNotFoundException(spec)
+            .firstOrNull { host -> host.isHostForSpec(spec) }
```

Comment:
I just noticed this. This could be expensive due to the fact that every invocation is an RPC call between Allocator and the Host via Intent on Android. At the very least, we will need to optimize this later with some form of caching and invalidation.

Reply:
Yup. We could also separate host descriptors from hosts if necessary (i.e. local objects in the allocator makes these decisions rather than the actual host objects).

I had to remove the getCanonicalName call from here because this is platform-independent code; going beyond that, a canonicalName check doesn't make sense in the general case as some particles won't even be implemented in Java/Kotlin.

Cameron's suggestion was to include a uniqueName in the particle schema and statically map those to hosts somehow. This seems like it would mostly work but there are issues around redirecting from Prod to Dev.